src/pages/guides/go/realtime-messaging.mdx

```diff
@@ -0,0 +1,231 @@
+
+```bash
+nitric start
+```
+
+If everything is working as expected, you can now delete all files/folders in the `services/` folder. We'll create new services in this guide.
+
+## Building the WebSocket Application
+
+Let's begin by setting up the WebSocket application. First, create a new folder called `websockets` within the services directory. Inside this folder, add a file named `main.go`, and include the following code:
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/nitrictech/go-sdk/handler"
+	"github.com/nitrictech/go-sdk/nitric"
+)
+
+func main() {
+  ws, err := nitric.NewWebsocket("public")
+  if err != nil {
+    fmt.Println("Error creating WebSocket:", err)
+    return
+  }
+
+  connections, err := nitric.NewKv("connections").Allow(nitric.KvStoreGet, nitric.KvStoreSet, nitric.KvStoreDelete)
+  if err != nil {
+    fmt.Println("Error creating KV store:", err)
+    return
+  }
+
+  // Add event handlers here
+
+	if err := nitric.Run(); err != nil {
+		fmt.Println("Error running Nitric service:", err)
+	}
+}
+```
+
+Here we're creating:
+
+- A WebSocket endpoint named `public`
+- A Key-Value store named `connections` to track WebSocket connections
+
+From here, let's add some features to that function that allow us to manage connections and broadcast messages.
+
+<Note>
+  You could separate some or all of these event handlers into their own services
+  if you prefer. For simplicity, we'll group them together in this guide.
+</Note>
+
+### Register connections on connect
+
+```go
+ws.On(handler.WebsocketConnect, func(ctx *handler.WebsocketContext, next handler.WebsocketHandler) (*handler.WebsocketContext, error) {
+  err := connections.Set(context.TODO(), ctx.Request.ConnectionID(), map[string]interface{}{
+    "connectionId": ctx.Request.ConnectionID(),
+  })
+  if err != nil {
+    return ctx, err
+  }
+
+  return next(ctx)
+})
+```
+
+### Remove connections on disconnect
+
+```go
+ws.On(handler.WebsocketDisconnect, func(ctx *handler.WebsocketContext, next handler.WebsocketHandler) (*handler.WebsocketContext, error) {
+  err := connections.Delete(context.TODO(), ctx.Request.ConnectionID())
+  if err != nil {
+    return ctx, err
+  }
+
+  return next(ctx)
+})
+```
+
+### Broadcast messages to all connected clients
+
+```go
+ws.On(handler.WebsocketMessage, func(ctx *handler.WebsocketContext, next handler.WebsocketHandler) (*handler.WebsocketContext, error) {
+  connectionStream, err := connections.Keys(context.TODO())
+  if err != nil {
+    return ctx, err
+  }
+
+  senderId := ctx.Request.ConnectionID()
+
+  for {
+    connectionId, err := connectionStream.Recv()
+    if err != nil {
+      break
+    }
+
+    if connectionId == senderId {
+      continue
+    }
+
+    message := fmt.Sprintf("%s: %s", senderId, ctx.Request.Message())
+    err = ws.Send(context.TODO(), connectionId, []byte(message))
+    if err != nil {
+      return ctx, err
+    }
+  }
+
+  return next(ctx)
+})
+```
```

Comment:
Might be helpful to link to a complete example, or provide the full example in a `<detail>` element that users can example to copy the whole thing in one go.

Reply:
Sure, added a project to the examples repo - will update a link here when that is merged